### PR TITLE
Assignment 1

### PR DESCRIPTION
**Pull request by tyheise**
We implemented and tested all features requested in Assignment 1.

- Feature 1: there is currently no check if there is a tie in the game (i.e., there is no longer a chance for
either player X or O to win). Implement such functionality, where after a move, a message will notify the
player if the game is in a tie and will end the game.
- Feature 2: add a scoreboard that keeps track of how many games were won, lost, tied by each player.
- Feature 3: the game currently ends whenever a player wins. Change it so it doesn't terminate right away,
but instead prompts the user whether they want to play another game or not. If yes, a new game will be
started. If not, the program will display the current score board and then terminate.

Enjoy!

### CODE REVIEW

File: src/main/java/cs/ualberta/cmput402/tictactoe/TicTacToeGame.java [hunk 1]
```diff
@@ -31,24 +31,46 @@ public void promptNextPlayer(){
...
+            System.out.println("Would you like to play again [y/n]?");
             String line = keyboardScanner.nextLine();
-            String input[] = line.split(",");
-            try {
-                board.playMove(Integer.parseInt(input[0]), Integer.parseInt(input[1]));
-            } catch (InvalidMoveException e) {
-                System.out.println("Invalid coordinates. Try again");
-                promptNextPlayer();
+
+            if(line.toLowerCase().equals("y")){
```
Comment (Perplex) [inline src/main/java/cs/ualberta/cmput402/tictactoe/TicTacToeGame.java]: You could use line.equalsIgnoreCase("y")

File: src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java [hunk 2]
```diff
@@ -116,7 +131,32 @@ public void printBoard(){
...
+    }
+
+    public void printScoreBoard() {
+
+	  	  System.out.println("Score Board");
+	      System.out.println("----------");
+	      System.out.println("X has won: " + scoreX + " times, and lost: " + scoreY + " times");
+	      System.out.println("O has won: " + scoreY + " times, and lost: " + scoreX + " times");
+          System.out.println("Number of ties: " + ties );
+	      System.out.println("----------");
```
Comment (Perplex) [inline src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java]: Indentation issues

File: src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java [hunk 3]
```diff
@@ -12,11 +14,18 @@
     private Player winner;
     private Player board[][];
 
+    public int scoreX = 0;
+    public int scoreY = 0;
```
Comment (Perplex) [inline src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java]: Use scoreO to keep naming uniform.

**Review by Perplex — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 4]
```diff
@@ -18,5 +18,23 @@
         </dependency>
 
     </dependencies>
+
+<build>
```
Comment (BlazPocrnja) [inline pom.xml]: `<build>...</build>` should be indented. `<plugins>...</plugins>` should be outdented.

File: pom.xml [hunk 5]
```diff
@@ -18,5 +18,23 @@
...
+                            <mainClass>cs.ualberta.cmput402.tictactoe.TicTacToeGame</mainClass>
+                        </manifest>
+                    </archive>
+                    <descriptorRefs>
+                        <descriptorRef>jar-with-dependencies</descriptorRef>
+                    </descriptorRefs>
+                </configuration>
+            </plugin>
+        </plugins>
+</build>
```
Comment (BlazPocrnja) [inline pom.xml]: `<build>`
...
`</build>`
Should be indented.

`<plugins>`
...
`</plugins>`
Should be outdented.

File: src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java [hunk 6]
```diff
@@ -128,5 +168,33 @@ public Player getPlayerAtPos(int row, int col){
...
+            for(int j = 0; j < 3; j++) {
+                if (board[i][j] == Player.NONE) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    private void incrementTies() {
```
Comment (BlazPocrnja) [inline src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java]: Unnecessary function. This is used only once and is a single line of code.

File: src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java [hunk 7]
```diff
@@ -116,7 +131,32 @@ public void printBoard(){
         }
     }
 
-    public Player getCurrentPlayer() {
+    public void scoreBoard(){
```
Comment (BlazPocrnja) [inline src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java]: Needs a more descriptive name. Something like "updateScore".

**Review by BlazPocrnja — CHANGES_REQUESTED** [review]

File: src/main/java/cs/ualberta/cmput402/tictactoe/TicTacToeGame.java [hunk 8]
```diff
@@ -31,24 +31,46 @@ public void promptNextPlayer(){
...
         Scanner keyboardScanner = new Scanner(System.in);
+        while (true){
+            while (board.getWinner() == null && !(board.isTie())){
+                board.printBoard();
+                promptNextPlayer();
+                String line = keyboardScanner.nextLine();
+                String input[] = line.split(",");
+                try {
+                    board.playMove(Integer.parseInt(input[0]), Integer.parseInt(input[1]));
+                } catch (InvalidMoveException e) {
```
Comment (SethBergen) [inline src/main/java/cs/ualberta/cmput402/tictactoe/TicTacToeGame.java]: Inconsistent use of whitespace surrounding brackets

**Review by SethBergen — COMMENTED** [review]
Minor style problems

File: src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java (hunk 7 above)
Comment (BlazPocrnja) [reply src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java]: I see @pennyfea has fixed this. Can you leave comments on the requested changes when you have completed them? Thanks!

File: src/main/java/cs/ualberta/cmput402/tictactoe/TicTacToeGame.java (hunk 1 above)
Comment (tyheise) [reply src/main/java/cs/ualberta/cmput402/tictactoe/TicTacToeGame.java]: I changed the method, I didn't know about that one!

File: pom.xml (hunk 5 above)
Comment (tyheise) [reply pom.xml]: Fixed

File: src/main/java/cs/ualberta/cmput402/tictactoe/TicTacToeGame.java (hunk 8 above)
Comment (tyheise) [reply src/main/java/cs/ualberta/cmput402/tictactoe/TicTacToeGame.java]: Made these consistent.

**Review by Perplex — DISMISSED** [review]

File: src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java (hunk 7 above)
Comment (pennyfea) [reply src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java]: Thanks for the comments. I made the requested changes. Names are now more descriptive.

Comment (pennyfea) [reply src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java]: Resolved  by  [ed240d7](https://github.com/Perplex/tictactoe/commit/ed240d7eff9f7396828639c06fc2c889eb71482b)

File: src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java (hunk 6 above)
Comment (ry-prz) [reply src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java]: Refactored in 82833d6 to get rid of this unnecessary function.

File: src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java (hunk 2 above)
Comment (ry-prz) [reply src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java]: Fixed these indentation inconsistencies in 9f96f5f

**Review by BlazPocrnja — APPROVED** [review]

File: src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java (hunk 3 above)
Comment (pennyfea) [reply src/main/java/cs/ualberta/cmput402/tictactoe/board/Board.java]: Fixed naming inconsistency in 6291399

**Review by Perplex — APPROVED** [review]

**Review by SethBergen — APPROVED** [review]

**Pull request merged**